crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -55,6 +58,13 @@ def checks
             passed: clusters_healthy?,
             errors: clusters_health_check_errors
           } if Api::Crowbar.addons.include?("ha")
+
+          errors = ret.select { |_k, v| v[:required] && v[:errors].any? }.map { |_k, v| v[:errors] }
+          if errors.any?
```

Comment:
Couldn't you just replace this whole `if` block with this?

    `upgrade_status.end_step(errors.any?, errors)`

And if so, to me that highlights the redundancy in the way the interface and state structures have been designed: they track both a list of errors, and whether there were any errors.  This needlessly complicates the code and also opens up the possibility of internal inconsistency, where these two values don't match (e.g. `success=false` and `errors={}` or vice-versa).